app/test-regex/main.c

```diff
@@ -0,0 +1,285 @@
+
+	mbuf_mp = rte_pktmbuf_pool_create("mbuf_pool",
+					TOTAL_MBUF_NUM, MBUF_CACHE_SIZE,
+					0, MBUF_SIZE,
+					rte_socket_id());
+	if (mbuf_mp == NULL)
+		rte_exit(EXIT_FAILURE, "Error: can't init mbuf pool\n");
+	rules_len = read_file(rules_file, &rules);
+	if (rules_len < 0)
+		rte_exit(EXIT_FAILURE, "Error: can't read rule file\n");
+	for (id = 0; id < num_devs; id++) {
+		ret = rte_regexdev_info_get(id, &info);
+		if (ret != 0)
+			rte_exit(EXIT_FAILURE,
+				"Error during getting device"
+				" (port %u) info: %s\n", id, strerror(-ret));
+		printf(":: initializing dev: %d\n", id);
+		nb_max_matches = info.max_matches;
+		nb_max_payload = info.max_payload_size;
+		if (info.regexdev_capa & RTE_REGEXDEV_SUPP_MATCH_AS_END_F)
+			dev_conf.dev_cfg_flags |= RTE_REGEXDEV_CFG_MATCH_AS_END_F;
+		dev_conf.nb_max_matches = info.max_matches;
+		dev_conf.nb_rules_per_group = info.max_rules_per_group;
+		dev_conf.rule_db_len = rules_len;
+		dev_conf.rule_db = rules;
+		ret = rte_regexdev_configure(id, &dev_conf);
+		if (ret < 0)
+			rte_exit(EXIT_FAILURE,
+				":: cannot configure device: err=%d, dev=%u\n",
+				ret, id);
+		ret = rte_regexdev_queue_pair_setup(id, 0, &qp_conf);
+		if (ret < 0)
+			rte_exit(EXIT_FAILURE,
+				":: cannot configure qp %d: err=%d, dev=%u\n",
+				0, ret, id);
+		printf(":: initializing device: %d done\n", id);
+	}
+}
+
+static void
+extbuf_free_cb(void *addr __rte_unused, void *fcb_opaque __rte_unused)
+{
+
+}
+
+#define START_BURST_SIZE 32
+
+static void
+main_loop(void)
+{
+	char *buf;
+	long buf_len;
+	long job_len;
+	uint32_t i;
+	struct rte_regex_ops *ops[TOTAL_MBUF_NUM];
+	uint16_t dev_id = 0;
+	uint16_t qp_id = 0;
+	uint8_t nb_matches;
+	struct rte_regexdev_match *match;
+	long pos;
+	uint32_t burst_size = RTE_MIN(START_BURST_SIZE, jobs);
+	unsigned long d_ind = 0;
+	struct rte_mbuf_ext_shared_info shinfo;
+	uint32_t dequeue = 0;
+	uint32_t enqueue;
+
+	shinfo.free_cb = extbuf_free_cb;
+	for (i = 0; i < TOTAL_MBUF_NUM; i++) {
+		ops[i] = rte_malloc(NULL, sizeof(*ops[0]) + nb_max_matches *
+				    sizeof(struct rte_regexdev_match), 0);
+		if (!ops[i])
+			goto main_loop_err;
+		ops[i]->mbuf = rte_pktmbuf_alloc(mbuf_mp);
+		if (!ops[i]->mbuf)
+			goto main_loop_err;
+	}
+
+	buf_len = read_file(data_file, &buf);
+	if (buf_len <= 0)
+		return;
+
+	job_len = buf_len / jobs;
+	if (job_len > nb_max_payload)
+		job_len = nb_max_payload;
+	pos = 0;
+
+	for (i = 0; (pos < buf_len) && (i < burst_size); i++) {
+		job_len = RTE_MIN(job_len, buf_len - pos);
+		rte_pktmbuf_attach_extbuf(ops[i]->mbuf, &buf[pos], 0, job_len,
+					  &shinfo);
+		ops[i]->mbuf->data_len = job_len;
+		ops[i]->mbuf->pkt_len = job_len;
+		ops[i]->user_id = i;
+		ops[i]->group_id0 = 1;
+		pos += job_len;
+	}
+	pos = 0;
+	while (pos < buf_len) {
+		enqueue = rte_regexdev_enqueue_burst(dev_id, qp_id,
+						     ops,
+						     burst_size);
+		
+		while (dequeue != enqueue)
+			dequeue += rte_regexdev_dequeue_burst(dev_id, qp_id, ops , enqueue);
```

Comment:
why force full completion and not send when we have space?
also this is in correct, the second read will destroy the result for the first read